File: crypto/password_validation.go
package crypto

import (
	_ "embed"
	"encoding/json"
	"fmt"
	"strings"
	"sync"
)

//go:embed password-requirements.json
var embeddedPasswordRequirements []byte

// PasswordRequirements holds password validation configuration
type PasswordRequirements struct {
	MinAccountPasswordLength    int    `json:"minAccountPasswordLength"`
	MinCustomPasswordLength     int    `json:"minCustomPasswordLength"`
	MinSharePasswordLength      int    `json:"minSharePasswordLength"`
	MaxPasswordLength           int    `json:"maxPasswordLength"`
	MinCharacterClassesRequired int    `json:"minCharacterClassesRequired"`
	SpecialCharacters           string `json:"specialCharacters"`
}

var (
	passwordRequirements     *PasswordRequirements
	passwordRequirementsOnce sync.Once
	passwordRequirementsErr  error
)

// LoadPasswordRequirements loads password requirements from embedded config
// The JSON is embedded at build time from crypto/password-requirements.json.
func LoadPasswordRequirements() (*PasswordRequirements, error) {
	passwordRequirementsOnce.Do(func() {
		passwordRequirements = &PasswordRequirements{}
		if err := json.Unmarshal(embeddedPasswordRequirements, passwordRequirements); err != nil {
			passwordRequirementsErr = fmt.Errorf("failed to parse embedded password requirements: %w", err)
			return
		}
	})

	return passwordRequirements, passwordRequirementsErr
}

// GetPasswordRequirements returns the loaded password requirements (panics if not loaded)
func GetPasswordRequirements() *PasswordRequirements {
	reqs, err := LoadPasswordRequirements()
	if err != nil {
		panic(fmt.Sprintf("Failed to load password requirements: %v", err))
	}
	return reqs
}

// GetEmbeddedPasswordRequirementsJSON returns the raw embedded JSON for API serving
func GetEmbeddedPasswordRequirementsJSON() []byte {
	return embeddedPasswordRequirements
}

// PasswordValidationResult represents the result of password validation
type PasswordValidationResult struct {
	MeetsRequirement bool              `json:"meets_requirements"`
	Requirements     RequirementChecks `json:"requirements"`
	Reasons          []string          `json:"reasons,omitempty"`
}

// RequirementChecks tracks individual password requirements
type RequirementChecks struct {
	Length          RequirementStatus `json:"length"`
	Uppercase       RequirementStatus `json:"uppercase"`
	Lowercase       RequirementStatus `json:"lowercase"`
	Number          RequirementStatus `json:"number"`
	Special         RequirementStatus `json:"special"`
	ClassCount      int               `json:"class_count"`
	ClassesRequired int               `json:"classes_required"`
}

// RequirementStatus represents the status of a single requirement
type RequirementStatus struct {
	Met     bool   `json:"met"`
	Current int    `json:"current,omitempty"`
	Needed  int    `json:"needed,omitempty"`
	Message string `json:"message"`
}

// ValidatePassword performs deterministic password validation.
// Pass = (length >= minLength) AND (length <= maxLength) AND (character classes met >= minCharacterClassesRequired)
// maxLength of 0 means no maximum is enforced.
func ValidatePassword(password string, minLength int, maxLength int, minClasses int, specialChars string) *PasswordValidationResult {
	length := len(password)

	// Check max length first (fail fast on absurdly long inputs)
	if maxLength > 0 && length > maxLength {
		return &PasswordValidationResult{
			MeetsRequirement: false,
			Requirements: RequirementChecks{
				Length: RequirementStatus{
					Met:     false,
					Current: length,
					Needed:  minLength,
					Message: fmt.Sprintf("Password too long (maximum %d characters)", maxLength),
				},
				Uppercase:       RequirementStatus{Met: false, Message: ""},
				Lowercase:       RequirementStatus{Met: false, Message: ""},
				Number:          RequirementStatus{Met: false, Message: ""},
				Special:         RequirementStatus{Met: false, Message: ""},
				ClassCount:      0,
				ClassesRequired: minClasses,
			},
			Reasons: []string{fmt.Sprintf("Password too long: %d characters (maximum %d)", length, maxLength)},
		}
	}

	hasUpper := false
	hasLower := false
	hasNumber := false
	hasSpecial := false

	for _, char := range password {
		switch {
		case char >= 'A' && char <= 'Z':
			hasUpper = true
		case char >= 'a' && char <= 'z':
			hasLower = true
		case char >= '0' && char <= '9':
			hasNumber = true
		default:
			if strings.ContainsRune(specialChars, char) {
				hasSpecial = true
			}
		}
	}

	classCount := 0
	if hasUpper {
		classCount++
	}
	if hasLower {
		classCount++
	}
	if hasNumber {
		classCount++
	}
	if hasSpecial {
		classCount++
	}

	lengthOK := length >= minLength
	classesOK := classCount >= minClasses
	meetsRequirement := lengthOK && classesOK

	// Build requirement checks
	checks := RequirementChecks{
		Length: RequirementStatus{
			Met:     lengthOK,
			Current: length,
			Needed:  minLength,
		},
		Uppercase:       RequirementStatus{Met: hasUpper},
		Lowercase:       RequirementStatus{Met: hasLower},
		Number:          RequirementStatus{Met: hasNumber},
		Special:         RequirementStatus{Met: hasSpecial},
		ClassCount:      classCount,
		ClassesRequired: minClasses,
	}

	// Set messages
	if checks.Length.Met {
		checks.Length.Message = fmt.Sprintf("Length requirement met (%d+ characters)", minLength)
	} else {
		remaining := minLength - length
		checks.Length.Message = fmt.Sprintf("Add %d more characters (currently %d/%d)", remaining, length, minLength)
	}

	if checks.Uppercase.Met {
		checks.Uppercase.Message = "Uppercase letter present"
	} else {
		checks.Uppercase.Message = "Missing: uppercase letter (A-Z)"
	}

	if checks.Lowercase.Met {
		checks.Lowercase.Message = "Lowercase letter present"
	} else {
		checks.Lowercase.Message = "Missing: lowercase letter (a-z)"
	}

	if checks.Number.Met {
		checks.Number.Message = "Number present"
	} else {
		checks.Number.Message = "Missing: number (0-9)"
	}

	if checks.Special.Met {
		checks.Special.Message = "Special character present"
	} else {
		checks.Special.Message = "Missing: special character"
	}

	// Build reasons for failure
	var reasons []string
	if !lengthOK {
		reasons = append(reasons, checks.Length.Message)
	}
	if !classesOK {
		// List which classes are missing
		var missing []string
		if !hasUpper {
			missing = append(missing, "uppercase (A-Z)")
		}
		if !hasLower {
			missing = append(missing, "lowercase (a-z)")
		}
		if !hasNumber {
			missing = append(missing, "number (0-9)")
		}
		if !hasSpecial {
			missing = append(missing, "special character")
		}
		reasons = append(reasons, fmt.Sprintf("Need %d character classes, have %d", minClasses, classCount))
		if len(missing) > 0 {
			reasons = append(reasons, fmt.Sprintf("Missing: %s", strings.Join(missing, ", ")))
		}
	}

	return &PasswordValidationResult{
		MeetsRequirement: meetsRequirement,
		Requirements:     checks,
		Reasons:          reasons,
	}
}

// ValidateAccountPassword validates account passwords using config requirements
func ValidateAccountPassword(password string) *PasswordValidationResult {
	reqs := GetPasswordRequirements()
	return ValidatePassword(password, reqs.MinAccountPasswordLength, reqs.MaxPasswordLength, reqs.MinCharacterClassesRequired, reqs.SpecialCharacters)
}

// ValidateSharePassword validates share passwords using config requirements
func ValidateSharePassword(password string) *PasswordValidationResult {
	reqs := GetPasswordRequirements()
	return ValidatePassword(password, reqs.MinSharePasswordLength, reqs.MaxPasswordLength, reqs.MinCharacterClassesRequired, reqs.SpecialCharacters)
}

// ValidateCustomPassword validates custom passwords using config requirements
func ValidateCustomPassword(password string) *PasswordValidationResult {
	reqs := GetPasswordRequirements()
	return ValidatePassword(password, reqs.MinCustomPasswordLength, reqs.MaxPasswordLength, reqs.MinCharacterClassesRequired, reqs.SpecialCharacters)
}

File: client/static/js/src/crypto/password-validation.ts
/**
 * Password Validation Module
 *
 * Deterministic password validation: length + character class checks.
 * Matches the Go backend (crypto/password_validation.go) exactly.
 *
 * Uses unified config from crypto/password-requirements.json served via API.
 * No zxcvbn, no entropy, no strength scores — pass/fail only.
 */

/**
 * Password requirements configuration (matches Go PasswordRequirements)
 */
interface PasswordConfig {
  minAccountPasswordLength: number;
  minCustomPasswordLength: number;
  minSharePasswordLength: number;
  minCharacterClassesRequired: number;
  specialCharacters: string;
}

// Password requirements - loaded from config file
let PASSWORD_CONFIG: PasswordConfig | null = null;

/**
 * Load password requirements from API endpoint.
 * Ensures client and server always use the same embedded configuration.
 */
async function loadPasswordConfig(): Promise<PasswordConfig> {
  if (PASSWORD_CONFIG !== null) {
    return PASSWORD_CONFIG;
  }

  try {
    const response = await fetch('/api/config/password-requirements');
    if (!response.ok) {
      throw new Error(`Failed to load password config: ${response.statusText}`);
    }
    const config: PasswordConfig = await response.json();
    PASSWORD_CONFIG = config;
    return config;
  } catch (error) {
    throw new Error(`Failed to load password requirements from API: ${error}`);
  }
}

/**
 * Requirement status for individual password checks
 */
export interface RequirementStatus {
  met: boolean;
  current?: number;
  needed?: number;
  message: string;
}

/**
 * Individual requirement checks (matches Go RequirementChecks)
 */
export interface RequirementChecks {
  length: RequirementStatus;
  uppercase: RequirementStatus;
  lowercase: RequirementStatus;
  number: RequirementStatus;
  special: RequirementStatus;
  class_count: number;
  classes_required: number;
}

/**
 * Password validation result (matches Go PasswordValidationResult)
 */
export interface PasswordValidationResult {
  meets_requirements: boolean;
  requirements: RequirementChecks;
  reasons: string[];
}

/**
 * Check password requirements deterministically.
 * Pass = (length >= minLength) AND (character classes >= minClasses)
 */
function checkPassword(
  password: string,
  minLength: number,
  minClasses: number,
  specialChars: string
): PasswordValidationResult {
  const length = password.length;
  let hasUpper = false;
  let hasLower = false;
  let hasNumber = false;
  let hasSpecial = false;

  for (const char of password) {
    const code = char.charCodeAt(0);
    if (code >= 65 && code <= 90) {
      hasUpper = true;
    } else if (code >= 97 && code <= 122) {
      hasLower = true;
    } else if (code >= 48 && code <= 57) {
      hasNumber = true;
    } else if (specialChars.includes(char)) {
      hasSpecial = true;
    }
  }

  let classCount = 0;
  if (hasUpper) classCount++;
  if (hasLower) classCount++;
  if (hasNumber) classCount++;
  if (hasSpecial) classCount++;

  const lengthOK = length >= minLength;
  const classesOK = classCount >= minClasses;
  const meetsRequirements = lengthOK && classesOK;

  // Build requirement checks
  const checks: RequirementChecks = {
    length: {
      met: lengthOK,
      current: length,
      needed: minLength,
      message: '',
    },
    uppercase: {
      met: hasUpper,
      message: '',
    },
    lowercase: {
      met: hasLower,
      message: '',
    },
    number: {
      met: hasNumber,
      message: '',
    },
    special: {
      met: hasSpecial,
      message: '',
    },
    class_count: classCount,
    classes_required: minClasses,
  };

  // Set messages
  if (checks.length.met) {
    checks.length.message = `Length requirement met (${minLength}+ characters)`;
  } else {
    const remaining = minLength - length;
    checks.length.message = `Add ${remaining} more character${remaining !== 1 ? 's' : ''} (currently ${length}/${minLength})`;
  }

  checks.uppercase.message = hasUpper
    ? 'Uppercase letter present'
    : 'Missing: uppercase letter (A-Z)';

  checks.lowercase.message = hasLower
    ? 'Lowercase letter present'
    : 'Missing: lowercase letter (a-z)';

  checks.number.message = hasNumber
    ? 'Number present'
    : 'Missing: number (0-9)';

  checks.special.message = hasSpecial
    ? 'Special character present'
    : 'Missing: special character';

  // Build failure reasons
  const reasons: string[] = [];
  if (!lengthOK) {
    reasons.push(checks.length.message);
  }
  if (!classesOK) {
    const missing: string[] = [];
    if (!hasUpper) missing.push('uppercase (A-Z)');
    if (!hasLower) missing.push('lowercase (a-z)');
    if (!hasNumber) missing.push('number (0-9)');
    if (!hasSpecial) missing.push('special character');
    reasons.push(`Need ${minClasses} character classes, have ${classCount}`);
    if (missing.length > 0) {
      reasons.push(`Missing: ${missing.join(', ')}`);
    }
  }

  return {
    meets_requirements: meetsRequirements,
    requirements: checks,
    reasons,
  };
}

/**
 * Validate password with explicit parameters
 */
export function validatePassword(
  password: string,
  minLength: number,
  minClasses: number,
  specialChars: string
): PasswordValidationResult {
  if (password === '') {
    return {
      meets_requirements: false,
      requirements: {
        length: { met: false, current: 0, needed: minLength, message: `Enter a password (minimum ${minLength} characters)` },
        uppercase: { met: false, message: 'Missing: uppercase letter (A-Z)' },
        lowercase: { met: false, message: 'Missing: lowercase letter (a-z)' },
        number: { met: false, message: 'Missing: number (0-9)' },
        special: { met: false, message: 'Missing: special character' },
        class_count: 0,
        classes_required: minClasses,
      },
      reasons: [`Enter a password (minimum ${minLength} characters)`],
    };
  }

  return checkPassword(password, minLength, minClasses, specialChars);
}

/**
 * Validate account password using config requirements
 */
export async function validateAccountPassword(
  password: string
): Promise<PasswordValidationResult> {
  const config = await loadPasswordConfig();
  return validatePassword(
    password,
    config.minAccountPasswordLength,
    config.minCharacterClassesRequired,
    config.specialCharacters
  );
}

/**
 * Validate share password using config requirements
 */
export async function validateSharePassword(
  password: string
): Promise<PasswordValidationResult> {
  const config = await loadPasswordConfig();
  return validatePassword(
    password,
    config.minSharePasswordLength,
    config.minCharacterClassesRequired,
    config.specialCharacters
  );
}

/**
 * Validate custom password using config requirements
 */
export async function validateCustomPassword(
  password: string
): Promise<PasswordValidationResult> {
  const config = await loadPasswordConfig();
  return validatePassword(
    password,
    config.minCustomPasswordLength,
    config.minCharacterClassesRequired,
    config.specialCharacters
  );
}

/**
 * Synchronous basic validation (for immediate feedback before async config loads).
 * Uses hardcoded defaults matching the JSON config.
 */
export function validatePasswordBasic(password: string, minLength: number): {
  meetsBasicRequirements: boolean;
  requirements: RequirementChecks;
} {
  // Use default special chars and 4 classes required as fallback
  const result = checkPassword(password, minLength, 4, '!@#$%^&*()_+-=[]{}|;:\'",.<>?/`~');
  return {
    meetsBasicRequirements: result.meets_requirements,
    requirements: result.requirements,
  };
}
